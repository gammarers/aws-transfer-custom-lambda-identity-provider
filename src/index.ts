import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as transfer from 'aws-cdk-lib/aws-transfer';
import { Construct } from 'constructs';
import crc from 'crc';

export interface TransferCustomLambdaIdentityProviderProps {
  readonly customHostname?: string;
  readonly route53HostedZoneId?: string;
}

export class TransferCustomLambdaIdentityProvider extends Construct {

  constructor(scope: Construct, id: string, props: TransferCustomLambdaIdentityProviderProps) {
    super(scope, id);

    const account = cdk.Stack.of(this).account;
    const region = cdk.Stack.of(this).region;

    const nameSuffix = crc.crc32(cdk.Names.uniqueId(this)).toString(16);

    // ðŸ‘‡Create Lambda Execution role.
    const lambdaExecutionRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `lambda-exec-role-${nameSuffix}`,
      description: '',
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        ['get-secret-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'secretsmanager:GetSecretValue',
              ],
              resources: [
                `arn:aws:secretsmanager:${region}:${account}:secret:transfer/s-*`,
              ],
            }),
          ],
        }),
      },
    });

    // ðŸ‘‡Create Lambda function.
    const lambdaFunction = new lambda.Function(this, 'Function', {
      functionName: `sftp-user-authentication-func-${nameSuffix}`,
      description: 'A function to lookup and return user data from AWS Secrets Manager.',
      runtime: lambda.Runtime.PYTHON_3_8,
      code: lambda.Code.fromAsset(path.join(__dirname, '../asset/sftp-user-authentication-func')),
      handler: 'index.lambda_handler',
      role: lambdaExecutionRole,
    });
    lambdaFunction.addPermission('LambdaAccessPermission', {
      principal: new iam.ServicePrincipal('transfer.amazonaws.com'),
    });

    // ðŸ‘‡Create Transfer Logging role.
    const loggingRole = new iam.Role(this, 'LoggingRole', {
      roleName: `transfer-logging-role-${nameSuffix}`,
      description: 'transfer logging role.',
      assumedBy: new iam.ServicePrincipal('transfer.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSTransferLoggingAccess'),
      ],
    });

    // ðŸ‘‡Create Transfer Server.
    const transferServer = new transfer.CfnServer(this, 'TransferServer', {
      protocols: ['SFTP'],
      endpointType: 'PUBLIC',
      identityProviderType: 'AWS_LAMBDA',
      identityProviderDetails: {
        function: lambdaFunction.functionArn,
      },
      loggingRole: loggingRole.roleArn,
    });
    if (props.customHostname) {
      cdk.Tags.of(transferServer).add('aws:transfer:customHostname', props.customHostname);
      if (props.route53HostedZoneId) {
        cdk.Tags.of(transferServer).add('aws:transfer:route53HostedZoneId', props.route53HostedZoneId);
      }
    } else {
      if (props.route53HostedZoneId) {
        throw new Error('Missing Custom Hostname. Is set HostZoneId, and set Route53 HostZoneId.');
      }
    }
  }
}
