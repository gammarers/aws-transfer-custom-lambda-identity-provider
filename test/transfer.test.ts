import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { TransferCustomLambdaIdentityProvider } from '../src';

describe('TransferCustomLambdaIdentityProvider Testing', () => {

  const app = new App();
  const account = '123456789012';
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: account,
      region: 'us-east-1',
    },
  });
  new TransferCustomLambdaIdentityProvider(stack, 'TransferCustomLambdaIdentityProvider', {
    customHostname: 'sftp.example.com',
    route53HostedZoneId: 'Z0000000000000000000Q',
  });

  const template = Template.fromStack(stack);

  it('Should has Lambda execution role.', () => {
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: Match.stringLikeRegexp('lambda-exec-role-'),
      AssumeRolePolicyDocument: Match.objectEquals({
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectEquals({
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          }),
        ]),
      }),
      ManagedPolicyArns: Match.arrayWith([
        {
          'Fn::Join': Match.arrayEquals([
            '',
            Match.arrayEquals([
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
            ]),
          ]),
        },
      ]),
    });
  });

  it('Should has Transfer logging role.', () => {
    template.hasResourceProperties('AWS::IAM::Role', {
      RoleName: Match.stringLikeRegexp('transfer-logging-role-'),
      AssumeRolePolicyDocument: Match.objectEquals({
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectEquals({
            Effect: 'Allow',
            Principal: {
              Service: 'transfer.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          }),
        ]),
      }),
      ManagedPolicyArns: Match.arrayWith([
        {
          'Fn::Join': Match.arrayEquals([
            '',
            Match.arrayEquals([
              'arn:',
              {
                Ref: 'AWS::Partition',
              },
              ':iam::aws:policy/service-role/AWSTransferLoggingAccess',
            ]),
          ]),
        },
      ]),
    });
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});

describe('TransferCustomLambdaIdentityProvider Error Testing', () => {

  const app = new App();
  const account = '123456789012';

  it('Should throw error (invalid).', () => {
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: account,
        region: 'us-east-1',
      },
    });
    const transferWrapper = () => {
      new TransferCustomLambdaIdentityProvider(stack, 'TransferCustomLambdaIdentityProvider', {
        route53HostedZoneId: 'Z0000000000000000000Q',
      });
    };
    expect(transferWrapper).toThrowError(Error);
  });
});