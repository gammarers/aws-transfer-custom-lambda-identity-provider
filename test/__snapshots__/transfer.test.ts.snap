// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferCustomLambdaIdentityProvider Testing Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TransferCustomLambdaIdentityProviderFunction6EA67D56": Object {
      "DependsOn": Array [
        "TransferCustomLambdaIdentityProviderLambdaExecutionRole6A614CFE",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "A function to lookup and return user data from AWS Secrets Manager.",
        "FunctionName": "sftp-user-authentication-func-6c4f6c0f",
        "Handler": "index.lambda_handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TransferCustomLambdaIdentityProviderLambdaExecutionRole6A614CFE",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TransferCustomLambdaIdentityProviderFunctionLambdaAccessPermission6B14587C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "TransferCustomLambdaIdentityProviderFunction6EA67D56",
            "Arn",
          ],
        },
        "Principal": "transfer.amazonaws.com",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TransferCustomLambdaIdentityProviderLambdaExecutionRole6A614CFE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": "arn:aws:secretsmanager:us-east-1:123456789012:secret:transfer/s-*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "get-secret-policy",
          },
        ],
        "RoleName": "lambda-exec-role-6c4f6c0f",
      },
      "Type": "AWS::IAM::Role",
    },
    "TransferCustomLambdaIdentityProviderLoggingRole6CBDB475": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "transfer.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "transfer logging role.",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSTransferLoggingAccess",
              ],
            ],
          },
        ],
        "RoleName": "transfer-logging-role-6c4f6c0f",
      },
      "Type": "AWS::IAM::Role",
    },
    "TransferCustomLambdaIdentityProviderTransferServer5ABD8CD4": Object {
      "Properties": Object {
        "EndpointType": "PUBLIC",
        "IdentityProviderDetails": Object {
          "Function": Object {
            "Fn::GetAtt": Array [
              "TransferCustomLambdaIdentityProviderFunction6EA67D56",
              "Arn",
            ],
          },
        },
        "IdentityProviderType": "AWS_LAMBDA",
        "LoggingRole": Object {
          "Fn::GetAtt": Array [
            "TransferCustomLambdaIdentityProviderLoggingRole6CBDB475",
            "Arn",
          ],
        },
        "Protocols": Array [
          "SFTP",
        ],
        "Tags": Array [
          Object {
            "Key": "aws:transfer:customHostname",
            "Value": "sftp.example.com",
          },
          Object {
            "Key": "aws:transfer:route53HostedZoneId",
            "Value": "Z0000000000000000000Q",
          },
        ],
      },
      "Type": "AWS::Transfer::Server",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
